presto-main/src/main/java/com/facebook/presto/execution/scheduler/SectionExecutionFactory.java

```diff
@@ -403,16 +515,54 @@ else if (partitioningHandle.equals(SCALED_WRITER_DISTRIBUTION)) {
         }
     }
 
-    private Optional<Predicate<Node>> getNodePoolSelectionPredicate(StreamingSubPlan plan)
+    private String getTaskIdentifier(com.facebook.presto.execution.TaskId failedTask)
     {
-        if (!isEnableWorkerIsolation || plan.getFragment().getStageExecutionDescriptor().isStageGroupedExecution()) {
+        return new StringBuilder()
+                .append("S")
+                .append(failedTask.getStageExecutionId().getStageId().getId())
+                .append(".")
+                .append(failedTask.getStageExecutionId().getId())
+                .append(".")
+                .append(failedTask.getId())
+                .append(".")
+                .append(failedTask.getAttemptNumber())
+                .toString();
+    }
+
+    private Optional<Predicate<Node>> getNodePoolSelectionPredicate(StreamingSubPlan plan, Session session, PartitioningHandle partitioningHandle)
+    {
+        if (!isEnableWorkerIsolation) {
             //skipping node pool based selection for grouped execution
             return Optional.empty();
         }
-        NodePoolType workerPoolType = plan.getFragment().isLeaf() ? LEAF : INTERMEDIATE;
+        //error out grouped execution query to clear the noise
+        checkArgument(!plan.getFragment().getStageExecutionDescriptor().isStageGroupedExecution(), "Grouped execution not supported");
+        boolean hasLocalExchangeAtRoot = PlanFragment.containLocalExchange(plan.getFragment().getRoot());
+        boolean isLeafFragment = isLeafSourceFragment(plan, partitioningHandle);
+        NodePoolType workerPoolType = !isLeafFragment || hasLocalExchangeAtRoot ? INTERMEDIATE : LEAF;
+        //avoid splitting node for jmx connector
+        if (isLeafFragment && plan.getFragment().getRoot() instanceof TableScanNode && ((TableScanNode) plan.getFragment().getRoot()).getTable().getConnectorId().getCatalogName().equals("jmx")) {
+            return Optional.empty();
+        }
+        ConcurrentHashMap<PlanFragmentId, Session.Pair<NodePoolType, String>> fragmentToPoolTypeMapping = session.getFragmentToPoolTypeMapping();
+        Session.Pair<NodePoolType, String> nodePoolTypeStringPair = fragmentToPoolTypeMapping.get(plan.getFragment().getId());
+        if (nodePoolTypeStringPair != null && nodePoolTypeStringPair.getKey() != workerPoolType) {
+            log.error("Error in pool type evaluation, plan =%s,hasLocalExchangeAtRoot = %s, isLeaf=%s, plan in map =%s", plan.getFragment().getJsonRepresentation().orElse(null), hasLocalExchangeAtRoot, plan.getFragment().isLeaf(), nodePoolTypeStringPair.getValue());
+            throw new RuntimeException("Error in pool type evaluation");
+        }
+        else {
+            fragmentToPoolTypeMapping.put(plan.getFragment().getId(), new Session.Pair<>(workerPoolType, plan.getFragment().getJsonRepresentation().orElse("")));
```

Comment:
AGPQN when do we enter into the erorr situation.
Also why do we need the entire plan fragment json stuffed into the session